core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -787,8 +787,37 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
 
   final def left[R1 <: R, C]: ZIO[Either[R1, C], E, Either[A, C]] = self +++ ZIO.identity[C]
 
+  /**
+   * Returns a successful effect if the value is 'Left', or fails with the error f.
+   */
+  def leftOrFail[B, C, E1 >: E](f: E1)(implicit ev: A <:< Either[B, C]): ZIO[R, E1, B] =
+    self.flatMap(ev(_).fold[ZIO[R, E1, B]](ZIO.succeed(_), _ => ZIO.fail(f)))
+
+  /**
+   * Returns a successful effect if the value is 'Left', or fails with a [[NoSuchElementException]].
+   */
+  def leftOrFailException[B, C, E1 >: NoSuchElementException](
+    implicit ev: A <:< Either[B, C],
+    ev2: E <:< E1
+  ): ZIO[R, E1, B] = self.map(_.left.get).mapError(ev2)
```

Comment:
2 problems here - this is throwing in `map` and will cause a defect (rather than a typed exception), and projections are deprecated in 2.13 if I am not mistaken.

Can you fold/pattern match on the Either instead?

Reply:
I agree with both points. please see my comment above, I would love some help here if possible.

Reply:
I agree on both points. 

I am struggling with the fold/pattern matching on the '*exception' methods and opened this PR to get feedback on how to do this properly(please see my comment above).

would love some help here if possible :)

Reply:
Sorry, I take back what I said about the `E1` - the bounds are necessary to unify `NoSuchElementException` with the existing `E`.

This should work without `right.get`:
```scala
  /**
   * Returns a successful effect if the value is `Left`, or fails with a [[scala.NoSuchElementException]].
   */
  def leftOrFailException[B, C, E1 >: NoSuchElementException](
    implicit ev: A <:< Either[B, C],
    ev2: E <:< E1
  ): ZIO[R, E1, B] =
    self.foldM(
      e => ZIO.fail(ev2(e)),
      a =>
        ev(a).fold(
          ZIO.succeed(_),
          _ => ZIO.fail(new NoSuchElementException)
        )
    )
```